### PR TITLE
network: dhcp4: support semi-static route with Gateway=_dhcp4 with UseGateway=no

### LINKED ISSUE #20208: networkd: Ability to add route with Gateway=_dhcp4 without getting default route from DHCP
**Is your feature request related to a problem? Please describe.**
The case where you have a mobile device with wlan connectivity where you want all of your traffic to go through a VPN is currently served pretty badly by systemd-networkd.
The want is to create a route from the gateway of any wlan you connect to, to the VPN endpoint and then to use another systemd unit to establish the VPN connection and default route.

Currently if `UseGateway=no` no routes referencing to gateway are added. This gets rid of the `default via @GATEWAY@ dev wlan0` route but in this case the `[Route]` with `Gateway=_dhcp4` is not added either.

The `UseRoutes=no` option does not seem to affect this.

**Describe the solution you'd like**
Ability to use `Gateway=_dhcp4` in routes without adding default route via gateway.

**Describe alternatives you've considered**
- Adding a "dummy" route with higher priority so that the default route by networkd does not allow any traffic to leak past. This is a hack and not a clean solution.
- Use another DHCP client with ability to hook into events. For example add a dummy route, wait for wlan to connect and to acquire DHCP lease, delete DHCP default route, delete dummy default route. And add the VPN default route at some point in that chain. This again is a horrible hack.

**The systemd version you checked that didn't have the feature you are asking for**
249 (Arch Linux version 249-2)

### PR DESCRIPTION
**Pull request by yuwata**
Closes #20208.

### CODE REVIEW

**Comment by bluca** [comment]
```
18:43:28 ======================================================================
18:43:28 FAIL: test_dhcp_client_ipv4_use_routes_gateway (__main__.NetworkdDHCPClientTests) (classless=False, dns_and_ntp_routes=True, gateway=False, routes=True)
18:43:28 ----------------------------------------------------------------------
18:43:28 Traceback (most recent call last):
18:43:28   File "./test/test-network/systemd-networkd-tests.py", line 3947, in test_dhcp_client_ipv4_use_routes_gateway
18:43:28     self._test_dhcp_client_ipv4_use_routes_gateway(routes, gateway, dns_and_ntp_routes, classless)
18:43:28   File "./test/test-network/systemd-networkd-tests.py", line 3995, in _test_dhcp_client_ipv4_use_routes_gateway
18:43:28     self.assertNotRegex(output, r'192.168.5.1 proto dhcp scope link src 192.168.5.[0-9]* metric 1024')
18:43:28 AssertionError: Regex matched: '192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024' matches '192.168.5.1 proto dhcp scope link src 192.168.5.[0-9]* metric 1024' in '8.0.0.0/8 via 192.168.5.3 proto dhcp src 192.168.5.181 metric 1024 \n8.8.8.8 via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n9.9.9.9 via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n192.168.5.0/24 proto kernel scope link src 192.168.5.181 metric 1024 \n192.168.5.0/24 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.3 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.10 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.11 proto dhcp scope link src 192.168.5.181 metric 1024'
18:43:28 
18:43:28 ======================================================================
18:43:28 FAIL: test_dhcp_client_ipv4_use_routes_gateway (__main__.NetworkdDHCPClientTests) (classless=True, dns_and_ntp_routes=True, gateway=False, routes=False)
18:43:28 ----------------------------------------------------------------------
18:43:28 Traceback (most recent call last):
18:43:28   File "./test/test-network/systemd-networkd-tests.py", line 3947, in test_dhcp_client_ipv4_use_routes_gateway
18:43:28     self._test_dhcp_client_ipv4_use_routes_gateway(routes, gateway, dns_and_ntp_routes, classless)
18:43:28   File "./test/test-network/systemd-networkd-tests.py", line 3995, in _test_dhcp_client_ipv4_use_routes_gateway
18:43:28     self.assertNotRegex(output, r'192.168.5.1 proto dhcp scope link src 192.168.5.[0-9]* metric 1024')
18:43:28 AssertionError: Regex matched: '192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024' matches '192.168.5.1 proto dhcp scope link src 192.168.5.[0-9]* metric 1024' in '8.8.8.8 via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n9.9.9.9 via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n192.168.5.0/24 proto kernel scope link src 192.168.5.181 metric 1024 \n192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.10 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.11 proto dhcp scope link src 192.168.5.181 metric 1024'
18:43:28 
18:43:28 ======================================================================
18:43:28 FAIL: test_dhcp_client_ipv4_use_routes_gateway (__main__.NetworkdDHCPClientTests) (classless=False, dns_and_ntp_routes=True, gateway=False, routes=False)
18:43:28 ----------------------------------------------------------------------
18:43:28 Traceback (most recent call last):
18:43:28   File "./test/test-network/systemd-networkd-tests.py", line 3947, in test_dhcp_client_ipv4_use_routes_gateway
18:43:28     self._test_dhcp_client_ipv4_use_routes_gateway(routes, gateway, dns_and_ntp_routes, classless)
18:43:28   File "./test/test-network/systemd-networkd-tests.py", line 3995, in _test_dhcp_client_ipv4_use_routes_gateway
18:43:28     self.assertNotRegex(output, r'192.168.5.1 proto dhcp scope link src 192.168.5.[0-9]* metric 1024')
18:43:28 AssertionError: Regex matched: '192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024' matches '192.168.5.1 proto dhcp scope link src 192.168.5.[0-9]* metric 1024' in '8.8.8.8 via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n9.9.9.9 via 192.168.5.1 proto dhcp src 192.168.5.181 metric 1024 \n192.168.5.0/24 proto kernel scope link src 192.168.5.181 metric 1024 \n192.168.5.1 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.10 proto dhcp scope link src 192.168.5.181 metric 1024 \n192.168.5.11 proto dhcp scope link src 192.168.5.181 metric 1024'
```

**Comment by yuwata** [comment]
Updated. Now, networkd also tries to find a default gateway in classless static routes even if `UseRoutes=no`.

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -406,6 +403,43 @@ static int dhcp4_request_static_routes(Link *link, struct in_addr *ret_default_g
...
+                                continue;
+
+                        r = sd_dhcp_route_get_gateway(static_routes[i], ret_default_gw);
+                        if (r < 0)
+                                return r;
+
+                        break;
+                }
+
+                return 0; /* Do not return 1 here, to make router option can override the found default gateway. */
```
Comment (bluca) [inline src/network/networkd-dhcp4.c]: I'd rephrase for extra clarity as "Do not return 1 here, to ensure the router option can override the default gateway that was found"

File: src/network/networkd-dhcp4.c [hunk 2]
```diff
@@ -454,19 +488,15 @@ static int dhcp4_request_static_routes(Link *link, struct in_addr *ret_default_g
         return classless_route;
 }
 
-static int dhcp4_request_gateway(Link *link, struct in_addr *ret_gw) {
+static int dhcp4_request_gateway(Link *link, struct in_addr *gw) {
```
Comment (bluca) [inline src/network/networkd-dhcp4.c:493]: why is the "ret_" prefix removed? Isn't it a return variable?

**Review by bluca — APPROVED** [review]
lgtm

File: src/network/networkd-dhcp4.c (hunk 2 above)
Comment (yuwata) [reply src/network/networkd-dhcp4.c:493]: We usually set "ret_" prefix when the value is unconditionally set on success. But, now this function sets conditionally. So, I dropped the prefix.

**Comment by yuwata** [comment]
@bluca Thank you for the review and suggestion. The comment in the code is updated. Upgrading the green label.
